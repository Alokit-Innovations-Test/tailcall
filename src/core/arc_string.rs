use std::sync::Arc;

use headers::HeaderValue;
use schemars::gen::SchemaGenerator;
use schemars::schema::Schema;
use serde::{Deserialize, Deserializer, Serialize, Serializer};

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ArcString(Arc<str>);

impl<'de> Deserialize<'de> for ArcString {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        String::deserialize(deserializer).map(ArcString::from)
    }
}

impl Serialize for ArcString {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.as_str().serialize(serializer)
    }
}

impl Default for ArcString {
    fn default() -> Self {
        ArcString("".into())
    }
}

impl AsRef<str> for ArcString {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

impl From<&str> for ArcString {
    fn from(s: &str) -> Self {
        ArcString(s.into())
    }
}

impl From<&String> for ArcString {
    fn from(s: &String) -> Self {
        ArcString(s.clone().into())
    }
}

impl From<String> for ArcString {
    fn from(s: String) -> Self {
        ArcString(s.into())
    }
}

impl From<ArcString> for Vec<u8> {
    fn from(s: ArcString) -> Self {
        s.as_str().as_bytes().to_vec()
    }
}

impl TryFrom<ArcString> for HeaderValue {
    type Error = <HeaderValue as TryFrom<String>>::Error;

    fn try_from(s: ArcString) -> Result<Self, Self::Error> {
        s.as_str().parse()
    }
}

impl PartialEq<&str> for ArcString {
    fn eq(&self, other: &&str) -> bool {
        self.as_ref() == *other
    }
}

impl schemars::JsonSchema for ArcString {
    fn schema_name() -> String {
        <String as schemars::JsonSchema>::schema_name()
    }

    fn json_schema(gen: &mut SchemaGenerator) -> Schema {
        <String as schemars::JsonSchema>::json_schema(gen)
    }
}

impl std::fmt::Display for ArcString {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        self.as_str().fmt(f)
    }
}

impl ArcString {
    pub fn as_str(&self) -> &str {
        self.as_ref()
    }
}
