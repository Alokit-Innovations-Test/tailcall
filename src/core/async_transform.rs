use std::future::Future;

use super::valid::{Valid, Validator};

/// A configuration transformer that allows us to perform various
/// transformations on the configuration before it's further processed for
/// blueprint creation.
pub trait AsyncTransform {
    type Value: Send;
    type Error;

    fn transform(
        &self,
        value: Self::Value,
    ) -> impl Future<Output = Valid<Self::Value, Self::Error>> + Send;
}

/// A suite of common operators that are available for all transformers.
pub trait AsyncTransformerOps: AsyncTransform + Send {
    fn pipe<Other: AsyncTransform<Value = Self::Value, Error = Self::Error>>(
        self,
        other: Other,
    ) -> Pipe<Self, Other>
    where
        Self: Sized,
    {
        Pipe(self, other)
    }

    fn generate(&self) -> impl Future<Output = Valid<Self::Value, Self::Error>> + Send
    where
        Self: Send + Sync,
        Self::Value: std::default::Default,
    {
        async move { self.transform(Self::Value::default()).await }
    }

    fn when(self, cond: bool) -> When<Self>
    where
        Self: Sized,
    {
        When(self, cond)
    }
}

impl<T: AsyncTransform + Sync + Send> AsyncTransformerOps for T {}

pub struct When<A>(A, bool);
impl<A: AsyncTransform + Sync + Send> AsyncTransform for When<A> {
    type Value = A::Value;
    type Error = A::Error;

    async fn transform(
        &self,
        value: Self::Value,
    ) -> Valid<Self::Value, Self::Error> {
        if self.1 {
            self.0.transform(value).await
        } else {
            Valid::succeed(value)
        }
    }
}

/// Represents a composition of two transformers.
pub struct Pipe<A, B>(A, B);

impl<A, E, X, Y> AsyncTransform for Pipe<X, Y>
where
    X: AsyncTransform<Value = A, Error = E> + Send + Sync,
    Y: AsyncTransform<Value = A, Error = E> + Send + Sync,
    A: Sync + Send,
    E: Sync + Send,
{
    type Value = A;
    type Error = E;

    async fn transform(
        &self,
        value: Self::Value,
    ) -> Valid<Self::Value, Self::Error> {
        let result = self.0.transform(value).await;
        match result.to_result() {
            Ok(result) => self.1.transform(result).await,
            Err(err) => Valid::from_validation_err(err),
        }
    }
}

/// Represents an empty transformer.
pub struct Default<A, E>(std::marker::PhantomData<(A, E)>);

impl<A: Send + Sync, E: Send + Sync> AsyncTransform for Default<A, E> {
    type Value = A;
    type Error = E;
    async fn transform(
        &self,
        value: Self::Value,
    ) -> Valid<Self::Value, Self::Error> { Valid::succeed(value) }
}

// pub fn default<A, E>() -> Default<A, E> {
//     Default(std::marker::PhantomData)
// }
