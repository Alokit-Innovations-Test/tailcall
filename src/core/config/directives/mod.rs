use lazy_static::lazy_static;

mod add_field;
mod alias;
mod cache;
mod call;
mod expr;
mod federation;
mod graphql;
mod grpc;
mod http;
mod js;
mod link;
mod modify;
mod omit;
mod protected;
mod server;
mod telemetry;
mod upstream;

pub use add_field::*;
pub use alias::*;
pub use cache::*;
pub use call::*;
pub use expr::*;
pub use federation::*;
pub use graphql::*;
pub use grpc::*;
pub use http::*;
pub use js::*;
pub use link::*;
pub use modify::*;
pub use omit::*;
pub use protected::*;
pub use server::*;
pub use telemetry::*;
pub use upstream::*;

use crate::core::directive::DirectiveCodec;

lazy_static! {
    pub static ref KNOWN_DIRECTIVES: Vec<String> = vec![
        add_field::AddField::directive_name(),
        alias::Alias::directive_name(),
        cache::Cache::directive_name(),
        call::Call::directive_name(),
        expr::Expr::directive_name(),
        graphql::GraphQL::directive_name(),
        grpc::Grpc::directive_name(),
        http::Http::directive_name(),
        js::JS::directive_name(),
        link::Link::directive_name(),
        modify::Modify::directive_name(),
        omit::Omit::directive_name(),
        protected::Protected::directive_name(),
        server::Server::directive_name(),
        telemetry::Telemetry::directive_name(),
        upstream::Upstream::directive_name(),
    ];
}
