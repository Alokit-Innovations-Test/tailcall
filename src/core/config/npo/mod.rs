mod identifier;
mod queries;

use std::{collections::HashSet, rc::Rc};

pub use identifier::*;
pub use queries::*;

use super::Config;

struct NPOIdentifier<'a> {
    config: &'a Config,
}

impl<'a> NPOIdentifier<'a> {
    pub fn new(config: &'a Config) -> NPOIdentifier {
        NPOIdentifier { config }
    }

    fn iter(
        &self,
        path: Chunk<FieldName<'a>>,
        type_name: TypeName<'a>,
        is_list: bool,
        visited: &mut HashSet<TypeName<'a>>,
    ) -> Chunk<Chunk<FieldName<'a>>> {
        if visited.contains(&type_name) {
            return Chunk::new();
        } else {
            visited.insert(type_name);
        }
        let mut chunks = Chunk::new();
        if let Some(type_of) = self.config.find_type(&type_name.as_str()) {
            for (name, field) in type_of.fields.iter() {
                let path = path.clone().append(FieldName::new(name));
                let is_batch = field.has_batched_resolver();

                if field.has_resolver() {
                    if !is_batch && is_list {
                        chunks = chunks.append(path.clone());
                    }
                }

                let is_list = is_list | field.list;

                chunks = chunks.concat(self.iter(
                    path,
                    TypeName::new(field.type_of.as_str()),
                    is_list,
                    visited,
                ))
            }
        }

        chunks
    }

    fn find_chunks(&self) -> Chunk<Chunk<FieldName<'a>>> {
        match &self.config.schema.query {
            None => Chunk::new(),
            Some(query) => self.iter(
                Chunk::new(),
                TypeName::new(query.as_str()),
                false,
                &mut HashSet::new(),
            ),
        }
    }

    fn find(&self) -> Vec<Vec<&'a str>> {
        self.find_chunks()
            .as_vec()
            .iter()
            .map(|chunk| {
                chunk
                    .as_vec()
                    .iter()
                    .map(|field_name| field_name.as_str())
                    .collect()
            })
            .collect()
    }
}

#[derive(Clone)]
enum Chunk<A> {
    Nil,
    Cons(A, Rc<Chunk<A>>),
    Concat(Rc<Chunk<A>>, Rc<Chunk<A>>),
}

impl<A> Chunk<A> {
    fn new() -> Self {
        Self::Nil
    }

    fn from_vec(vec: Vec<A>) -> Chunk<A> {
        let mut path = Chunk::new();
        for a in vec.into_iter().rev() {
            path = Chunk::Cons(a, Rc::new(path));
        }
        path
    }

    fn append(self, a: A) -> Self {
        Chunk::Cons(a, Rc::new(self))
    }

    fn concat(self, other: Chunk<A>) -> Self {
        Self::Concat(Rc::new(self), Rc::new(other))
    }

    fn as_vec(&self) -> Vec<&A> {
        let mut vec = Vec::new();
        self.as_vec_mut(&mut vec);
        vec
    }

    fn as_vec_mut<'a>(&'a self, buf: &mut Vec<&'a A>) {
        match self {
            Chunk::Nil => {}
            Chunk::Cons(a, rest) => {
                buf.push(a);
                rest.as_vec_mut(buf);
            }
            Chunk::Concat(a, b) => {
                a.as_vec_mut(buf);
                b.as_vec_mut(buf);
            }
        }
    }
}
