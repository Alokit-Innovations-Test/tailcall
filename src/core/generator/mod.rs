mod from_json;
pub mod config;
mod from_proto;
mod generator;
mod graphql_type;
mod json;
mod proto;
mod from_generalized_config;
pub mod source;

pub use from_json::{from_json, ConfigGenerationRequest};
pub use json::NameGenerator;
pub use generator::Generator;
pub use from_generalized_config::FromGeneralizedConfig;
