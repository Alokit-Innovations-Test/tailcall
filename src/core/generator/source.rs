use thiserror::Error;

///
/// A list of sources from which a configuration can be created
#[derive(Clone, Copy, PartialEq, Debug, Default)]
pub enum ImportSource {
    #[default]
    Proto,
}

impl ImportSource {
    fn ext(&self) -> &str {
        match self {
            ImportSource::Proto => "proto",
        }
    }

    fn ends_with(&self, file: &str) -> bool {
        file.ends_with(&format!(".{}", self.ext()))
    }

    /// Detect the config format from the file name
    pub fn detect(name: &str) -> Result<Self, UnsupportedFileFormat> {
        const ALL: &[ImportSource] = &[ImportSource::Proto];

        ALL.iter()
            .find(|format| format.ends_with(name))
            .copied()
            .ok_or(UnsupportedFileFormat(name.to_string()))
    }
}

#[derive(Clone, Copy, PartialEq, Debug)]
pub enum ConfigSource {
    Json,
    Yml,
}

impl ConfigSource {
    fn ext(&self) -> &str {
        match self {
            Self::Json => "json",
            Self::Yml => "yml",
        }
    }

    fn ends_with(&self, file: &str) -> bool {
        file.ends_with(&format!(".{}", self.ext()))
    }

    /// Detect the config format from the file name
    pub fn detect(name: &str) -> Result<Self, UnsupportedFileFormat> {
        const ALL: &[ConfigSource] = &[ConfigSource::Json, ConfigSource::Yml];

        ALL.iter()
            .find(|format| format.ends_with(name))
            .copied()
            .ok_or(UnsupportedFileFormat(name.to_string()))
    }
}

#[derive(Debug, Error, PartialEq)]
#[error("Unsupported config extension: {0}")]
pub struct UnsupportedFileFormat(String);

impl std::str::FromStr for ImportSource {
    type Err = UnsupportedFileFormat;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "proto" => Ok(ImportSource::Proto),
            _ => Err(UnsupportedFileFormat(s.to_string())),
        }
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use super::*;

    #[test]
    fn test_from_str() {
        assert_eq!(ImportSource::from_str("proto"), Ok(ImportSource::Proto));
        assert!(ImportSource::from_str("foo").is_err());
    }
}
