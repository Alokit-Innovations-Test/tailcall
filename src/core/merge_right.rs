use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
use std::sync::Arc;

use serde_yaml::Value;

use super::valid::{Valid, Validator};

pub trait MergeRight: Sized {
    fn merge_right(self, other: Self) -> Valid<Self, String>;
}

impl<A: MergeRight> MergeRight for Option<A> {
    fn merge_right(self, other: Self) -> Valid<Self, String> {
        let valid = match (self, other) {
            (Some(this), Some(that)) => return this.merge_right(that).map(Some),
            (None, Some(that)) => Some(that),
            (Some(this), None) => Some(this),
            (None, None) => None,
        };

        Valid::succeed(valid)
    }
}

impl<A: MergeRight + Default> MergeRight for Arc<A> {
    fn merge_right(self, other: Self) -> Valid<Self, String> {
        let l = Arc::into_inner(self);
        let r = Arc::into_inner(other);
        l.merge_right(r).map(|x| Arc::new(x.unwrap_or_default()))
    }
}

impl<A> MergeRight for Vec<A> {
    fn merge_right(mut self, other: Self) -> Valid<Self, String> {
        self.extend(other);
        Valid::succeed(self)
    }
}

impl<K, V> MergeRight for BTreeMap<K, V>
where
    K: Ord,
    V: Clone + MergeRight,
{
    fn merge_right(mut self, other: Self) -> Valid<Self, String> {
        Valid::from_iter(other, |(name, other_value)| match self.remove(&name) {
            Some(value) => value.merge_right(other_value).map(|value| (name, value)),
            None => Valid::succeed((name, other_value)),
        })
        .map(Self::from_iter)
    }
}

impl<V> MergeRight for BTreeSet<V>
where
    V: Ord,
{
    fn merge_right(mut self, other: Self) -> Valid<Self, String> {
        self.extend(other);
        Valid::succeed(self)
    }
}

impl<V> MergeRight for HashSet<V>
where
    V: Eq + std::hash::Hash,
{
    fn merge_right(mut self, other: Self) -> Valid<Self, String> {
        self.extend(other);
        Valid::succeed(self)
    }
}

impl<K, V> MergeRight for HashMap<K, V>
where
    K: Eq + std::hash::Hash,
{
    fn merge_right(mut self, other: Self) -> Valid<Self, String> {
        self.extend(other);
        Valid::succeed(self)
    }
}

impl MergeRight for Value {
    fn merge_right(self, other: Self) -> Valid<Self, String> {
        let value = match (self, other) {
            (Value::Null | Value::Bool(_) | Value::Number(_) | Value::String(_), other) => other,
            (Value::Sequence(mut lhs), other) => match other {
                Value::Sequence(rhs) => {
                    lhs.extend(rhs);
                    Value::Sequence(lhs)
                }
                other => {
                    lhs.push(other);
                    Value::Sequence(lhs)
                }
            },
            (Value::Mapping(mut lhs), other) => match other {
                Value::Mapping(rhs) => {
                    return Valid::from_iter(rhs, |(key, value)| match lhs.remove(&key) {
                        Some(lhs_value) => lhs_value.merge_right(value).map(|value| (key, value)),
                        None => Valid::succeed((key, value)),
                    })
                    .map(|it| Value::Mapping(serde_yaml::Mapping::from_iter(it)))
                }
                Value::Sequence(mut rhs) => {
                    rhs.push(Value::Mapping(lhs));
                    Value::Sequence(rhs)
                }
                other => other,
            },
            (Value::Tagged(lhs), other) => match other {
                Value::Tagged(rhs) => {
                    if lhs.tag == rhs.tag {
                        let tag = lhs.tag;

                        return lhs.value.merge_right(rhs.value).map(|value| {
                            Value::Tagged(Box::new(serde_yaml::value::TaggedValue { tag, value }))
                        });
                    } else {
                        Value::Tagged(rhs)
                    }
                }
                other => other,
            },
        };

        Valid::succeed(value)
    }
}
