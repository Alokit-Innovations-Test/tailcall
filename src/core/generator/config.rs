use std::path::Path;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::core::config::{self, reader::ConfigReader};

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub enum InputSource {
    /// src: maintains the same src written in config.
    /// resolved_src: holds the correctly resolved src with respect to config.
    Config {
        src: String,
        resolved_src: Option<String>,
    },
    Import {
        src: String,
        resolved_src: Option<String>,
    },
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Input {
    #[serde(flatten)]
    pub source: InputSource,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Output {
    /// Controls the output format (graphql, json, yaml)
    pub format: config::Source,
    /// Specifies the output file name
    pub file: String,
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Schema {
    #[serde(default = "defaults::schema::query")]
    pub query: String,
    #[serde(default = "defaults::schema::mutation")]
    pub mutation: String,
}

impl Default for Schema {
    fn default() -> Self {
        Self {
            query: defaults::schema::query(),
            mutation: defaults::schema::mutation(),
        }
    }
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GenerateOptions {
    #[serde(default)]
    pub schema: Schema,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Transform {
    // TODO: change types
    pub ambiguous_name_resolver: Option<serde_json::Value>,
    pub merge_types: Option<serde_json::Value>,
    pub js: Option<serde_json::Value>,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GeneratorConfig {
    pub input: Vec<Input>,
    #[serde(default)]
    pub output: Output,
    #[serde(default)]
    pub generate: GenerateOptions,
    #[serde(default)]
    pub transform: Transform,
}

impl GeneratorConfig {
    /// Resolves all the paths present inside the GeneratorConfig.
    pub fn resolve_paths(mut self, file_path: &str) -> Self {
        let config_path = Path::new(file_path).parent().unwrap_or(Path::new(""));

        for input in self.input.iter_mut() {
            match &mut input.source {
                InputSource::Config { src, resolved_src }
                | InputSource::Import { src, resolved_src } => {
                    *resolved_src = Some(ConfigReader::resolve_path(src, Some(config_path)));
                }
            }
        }

        self.output.file = ConfigReader::resolve_path(&self.output.file, Some(config_path));

        self
    }
}

mod defaults {
    pub mod schema {
        pub fn query() -> String {
            "Query".to_string()
        }

        pub fn mutation() -> String {
            "Mutation".to_string()
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::GeneratorConfig;

    #[test]
    fn test_from_json() {
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();

        assert_debug_snapshot!(&config);
    }

    #[test]
    fn test_resolve_paths() {
        let file_path = "../../../tailcall-fixtures/fixtures/generator/simple-json.json";
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();
        let config = config.resolve_paths(file_path);
        assert_debug_snapshot!(&config);
    }
}
