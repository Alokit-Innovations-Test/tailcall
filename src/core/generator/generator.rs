use anyhow::Result;
use futures_util::future::join_all;
use prost_reflect::prost_types::FileDescriptorSet;
use prost_reflect::DescriptorPool;

use crate::core::generator::from_proto::from_proto;
use crate::core::generator::source::ImportSource;
use crate::core::merge_right::MergeRight;
use crate::core::proto_reader::ProtoReader;
use crate::core::resource_reader::ResourceReader;
use crate::core::runtime::TargetRuntime;
use crate::core::{
    config::{self, Config, ConfigModule, Link, LinkType, Resolution},
    resource_reader::Cached,
};

use super::config::{GeneratorConfig, InputSource};

// this function resolves all the names to fully-qualified syntax in descriptors
// that is important for generation to work
// TODO: probably we can drop this in case the config_reader will use
// protox::compile instead of more low-level protox_parse::parse
fn resolve_file_descriptor_set(descriptor_set: FileDescriptorSet) -> Result<FileDescriptorSet> {
    let descriptor_set = DescriptorPool::from_file_descriptor_set(descriptor_set)?;
    let descriptor_set = FileDescriptorSet {
        file: descriptor_set
            .files()
            .map(|file| file.file_descriptor_proto().clone())
            .collect(),
    };

    Ok(descriptor_set)
}

pub struct Generator {
    runtime: TargetRuntime,
    reader: ResourceReader<Cached>,
    proto_reader: ProtoReader,
}

impl Generator {
    pub fn new(runtime: TargetRuntime) -> Self {
        let reader = ResourceReader::cached(runtime.clone());

        Self {
            runtime: runtime.clone(),
            reader: reader.clone(),
            proto_reader: ProtoReader::init(reader, runtime),
        }
    }

    pub async fn run(&self, gen_config: GeneratorConfig) -> Result<()> {
        let resolvers = gen_config.input.into_iter().map(|input| async {
            match input.source {
                InputSource::Config { src } => {
                    let source = config::Source::detect(&src)?;
                    let schema = self.reader.read_file(&src).await?;

                    Config::from_source(source, &schema.content)
                }
                InputSource::Import { src } => {
                    let source = ImportSource::detect(&src)?;

                    match source {
                        ImportSource::Proto => {
                            let metadata = self.proto_reader.read(&src).await?;
                            let descriptor_set =
                                resolve_file_descriptor_set(metadata.descriptor_set)?;
                            let mut config = from_proto(&[descriptor_set], "Query")?;

                            config.links.push(Link {
                                id: None,
                                src: metadata.path,
                                type_of: LinkType::Protobuf,
                            });

                            Ok(config)
                        }
                    }
                }
            }
        });

        let mut config = Config::default();
        for result in join_all(resolvers).await {
            config = config.merge_right(result?)
        }

        let config = ConfigModule::from(config).resolve_ambiguous_types(|v| Resolution {
            input: format!("{}Input", v),
            output: v.to_owned(),
        });

        let output_source = config::Source::detect(&gen_config.output.file)?;

        let config = match output_source {
            config::Source::Json => config.to_json(true)?,
            config::Source::Yml => config.to_yaml()?,
            config::Source::GraphQL => config.to_sdl(),
        };

        self.runtime
            .file
            .write(&gen_config.output.file, config.as_bytes())
            .await?;

        Ok(())
    }

    pub async fn read_all<T: AsRef<str>>(
        &self,
        input_source: ImportSource,
        files: &[T],
        query: &str,
    ) -> Result<ConfigModule> {
        let mut links = vec![];
        let proto_metadata = self.proto_reader.read_all(files).await?;

        let mut config = Config::default();
        for metadata in proto_metadata {
            match input_source {
                ImportSource::Proto => {
                    links.push(Link { id: None, src: metadata.path, type_of: LinkType::Protobuf });
                    let descriptor_set = resolve_file_descriptor_set(metadata.descriptor_set)?;
                    config = config.merge_right(from_proto(&[descriptor_set], query)?);
                }
            }
        }

        config.links = links;
        Ok(
            ConfigModule::from(config).resolve_ambiguous_types(|v| Resolution {
                input: format!("{}Input", v),
                output: v.to_owned(),
            }),
        )
    }
}

#[cfg(test)]
mod test {
    use std::path::PathBuf;

    use tailcall_fixtures::protobuf;

    use super::*;

    fn start_mock_server() -> httpmock::MockServer {
        httpmock::MockServer::start()
    }

    #[tokio::test]
    async fn test_read_all() {
        let server = start_mock_server();
        let runtime = crate::core::runtime::test::init(None);
        let test_dir = PathBuf::from(tailcall_fixtures::protobuf::SELF);

        let news_content = runtime.file.read(protobuf::NEWS).await.unwrap();
        let greetings_a = runtime.file.read(protobuf::GREETINGS_A).await.unwrap();

        server.mock(|when, then| {
            when.method(httpmock::Method::GET).path("/news.proto");
            then.status(200)
                .header("Content-Type", "application/vnd.google.protobuf")
                .body(&news_content);
        });

        server.mock(|when, then| {
            when.method(httpmock::Method::GET)
                .path("/greetings_a.proto");
            then.status(200)
                .header("Content-Type", "application/protobuf")
                .body(&greetings_a);
        });

        let generator = Generator::new(runtime);
        let news = format!("http://localhost:{}/news.proto", server.port());
        let greetings_a = format!("http://localhost:{}/greetings_a.proto", server.port());
        let greetings_b = test_dir
            .join("greetings_b.proto")
            .to_str()
            .unwrap()
            .to_string();

        let config = generator
            .read_all(ImportSource::Proto, &[news, greetings_a, greetings_b], "Query")
            .await
            .unwrap();

        assert_eq!(config.links.len(), 3);
        assert_eq!(config.types.get("Query").unwrap().fields.len(), 8);
    }
}
