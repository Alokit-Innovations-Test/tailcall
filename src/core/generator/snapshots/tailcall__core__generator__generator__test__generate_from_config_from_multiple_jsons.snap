---
source: src/core/generator/generator.rs
expression: cfg_module.config.to_sdl()
---
schema @server @upstream(baseURL: "https://example.com") {
  query: Query
}

scalar Any

type F2 {
  campaignTemplates: Any
  colors: [Any]
}

type F4 {
  adult: Boolean
  age: Int
  name: String
}

type Query {
  f2: F2 @http(path: "/")
  f3: [Any] @http(path: "/api/v2/users")
  f4: [F4] @http(path: "/users")
}
